### DISCUSSION TITLE
Requesting help with bad formatting for text with negative left side bearing

### DISCUSSION THREAD

**Discussion by bclever1** [harfbuzz/harfbuzz#3441] [Q&A] [answered]
Hello all, 

We are working on incorporating Harfbuzz version 3.3.2 into our label printing stack for better kerning. 
For our font files we are using a number of TTFs downloaded from Google Fonts  
(https://fonts.google.com/noto/specimen/Noto+Sans?query=noto+ is one such example). 
With this installation of Harfbuzz we aren't seeing characters with negative left side bearing (e.g. 'j' and 'J') being positioned correctly.
 
FontForge and the Dutch Type Library OTM Light tool both display the bearings, with FontForge adjusting the positions correctly.

To investigate this issue, we built the tutorial found here: https://github.com/harfbuzz/harfbuzz-tutorial 
and do not see the hb_glyph_position_t.x_offsets being populated. All locations in the Harfbuzz code at which theses values might get set are skipped.

We are wondering if there is an hb_feature we need to enable to achieve this correction, or a value we need to set on the hb_font or hb_face? Are the google fonts somehow lacking? Is it a bug in this version of Harfbuzz?   

Many thanks for your consideration.

**Comment by behdad** [comment, marked as answer]
Hi,

What you are seeing is correct behavior from HarfBuzz. The shaping positions those letters in a way that they will extrude. That's expected behavior.

**Reply by bclever1** [reply]
Thank you for your response behdad. We updated our harfbuzz package to version 3.4, and by applying the examples of blob, font and buffer creation/configuration we are getting excellent left-side bearing placement from harfbuzz. We are now looking into kerning issues. We are turning on the kern feature correctly, and see the gpos table being accessed but we aren't seeing any effect on our positioning. An example of incorrect kerning is shown by the text "BAH PAH KAH NAH". All the pairs "XA" have a negative kern value in the font file. Our result is attached here: 
![harfbuzz_kern](https://user-images.githubusercontent.com/34148260/155377799-88a494e4-caf1-4b17-85b1-e828266c7bdb.PNG)
 
We are using the Noto Sans font from Google: 
https://fonts.google.com/?preview.text=BAH%20KAK%20NAH&preview.text_type=custom&query=noto+sans

Can you please provide a hint for how we can achieve the desired result? 

Kindest regards,
Brian Clever

**Reply by behdad** [reply]
This should just work out of the box and you should see the glyph advances adjusted for kerning. How are you using HarfBuzz? How are you constructing a face? What font-funcs are you using? Are you setting the font-scale correctly?

**Reply by bclever1** [reply]
Many thanks for your attention behdad.
We have HarfBuzz built as a static lib for Windows 10, x86 platform.

Here is the code relevant for us gathering the advances. Except for the addition of the 'harfBuzzFeatures' and 'allChars' vectors this is the same code as in the hb_shape_fuzzer project in the HarfBuzz distribution:

	std::vector<glyphData>allChars;

	std::vector<hb_feature_t> harfBuzzFeatures;

	const hb_tag_t KernTag = HB_TAG('k', 'e', 'r', 'n'); // kerning operations
	hb_feature_t KerningOn = { KernTag, 1, 0, std::numeric_limits<unsigned int>::max() };

	harfBuzzFeatures.push_back(KerningOn);

	hb_blob_t* blob = hb_blob_create_from_file_or_fail("C:\\Program Files (x86)\\ProShip\\Server\\labeler\\fonts\\NotoSans-Regular.ttf");
	unsigned len = 0;
	const char *font_data = hb_blob_get_data(blob, &len);

	//hb_blob_t *blob = hb_blob_create((const char *)data, size, HB_MEMORY_MODE_READONLY, nullptr, nullptr);

	hb_face_t *face = hb_face_create(blob, 0);
	hb_font_t *font = hb_font_create(face);
	hb_ot_font_set_funcs(font);
	hb_font_set_scale(font, hb_fontScale, hb_fontScale);

	unsigned num_coords = 0;
	if (len) num_coords = font_data[len - 1];
	num_coords = hb_ot_var_get_axis_count(face) > num_coords ? num_coords : hb_ot_var_get_axis_count(face);
	int *coords = (int *)calloc(num_coords, sizeof(int));
	if (len > num_coords + 1)
		for (unsigned i = 0; i < num_coords; ++i)
			coords[i] = ((int)font_data[len - num_coords + i - 1] - 128) * 10;
	hb_font_set_var_coords_normalized(font, coords, num_coords);
	free(coords);

	char* text = (char*)inStr;
	hb_buffer_t *buffer = hb_buffer_create();
	hb_buffer_set_flags (buffer, HB_BUFFER_FLAG_VERIFY);
	hb_buffer_add_utf8(buffer, text, -1, 0, -1);
	hb_buffer_guess_segment_properties(buffer);
	hb_shape(font, buffer, &(harfBuzzFeatures[0]), 1);

	unsigned int bufflen = hb_buffer_get_length(buffer);
	hb_glyph_info_t *info = hb_buffer_get_glyph_infos(buffer, NULL);
	hb_glyph_position_t *pos = hb_buffer_get_glyph_positions(buffer, NULL);

	for (unsigned int i = 0; i < bufflen; i++)
	{
		hb_codepoint_t gid = info[i].codepoint;
		unsigned int cluster = info[i].cluster;
		double x_advance = pos[i].x_advance;

		allChars.push_back({ text[i], (uint32_t)text[i], x_advance });
	}

**Reply by bclever1** [reply]
Also, if there is a more appropriate forum for questions like this I'd be happy to use it.

**Reply by behdad** [reply]
Okay let's remove the unnecessary code here. That whole block with `hb_font_set_var_coords_normalized` should go. The manual kerning on also can go as that's the default.

You should also take glyph x/y offset into consideration. Other than that, this should work. What's `hb_fontScale` here? That points to the possible problem. That, and the fact that you are ignoring the glyph offsets.

**Comment by bclever1** [comment]
Thank you behdad, we appreciate your help. Here is the simplified code with your suggestions as above. The x_offsets are always zero. The 'hb_font_scale' value is for dynamic debug purposes, we range it over 12 to 72 with no change in the kerning or x_offsets. Is there a preferred target value? 


std::vector<glyphData>allChars;

	hb_blob_t* blob = hb_blob_create_from_file_or_fail("C:\\Program Files (x86)\\ProShip\\Server\\labeler\\fonts\\NotoSans-Regular.ttf");
	unsigned len = 0;
	const char *font_data = hb_blob_get_data(blob, &len);

	hb_face_t *face = hb_face_create(blob, 0);
	hb_font_t *font = hb_font_create(face);
	hb_ot_font_set_funcs(font);
	hb_font_set_scale(font, hb_fontScale, hb_fontScale);

	char* text = (char*)inStr;
	hb_buffer_t *buffer = hb_buffer_create();
	hb_buffer_set_flags (buffer, HB_BUFFER_FLAG_VERIFY);
	hb_buffer_add_utf8(buffer, text, -1, 0, -1);
	hb_buffer_guess_segment_properties(buffer);
	hb_shape(font, buffer, nullptr, 0);

	unsigned int bufflen = hb_buffer_get_length(buffer);
	hb_glyph_info_t *info = hb_buffer_get_glyph_infos(buffer, NULL);
	hb_glyph_position_t *pos = hb_buffer_get_glyph_positions(buffer, NULL);

	for (unsigned int i = 0; i < bufflen; i++)
	{
		hb_codepoint_t gid = info[i].codepoint;
		unsigned int cluster = info[i].cluster;
		double x_advance = pos[i].x_advance;
		double x_offset = pos[i].x_offset;

		if (x_offset != 0)
		{
			_asm nop;
		}

		allChars.push_back({ text[i], (uint32_t)text[i], x_advance });
	}

**Comment by behdad** [comment]
Okay. Let's debug this further. The `hb_ot_font_set_funcs()` is also optional these days. For the preferred scale, Some sub-pixel precision is preferred but that should not be the problem here.

So if you print the advances you are getting, then what do you get for 'P' and 'PA'?

I'm also interested in the exact version of NotoSans you are using.

**Reply by bclever1** [reply]
With the font scale at 64,

For "P" I get advances = {39}. For "PA" I get advances = {37, 41}.

I'm attaching the font file we are using... relatively newly downloaded from google fonts.
[NotoSans-Regular.zip](https://github.com/harfbuzz/harfbuzz/files/8135892/NotoSans-Regular.zip)

**Reply by behdad** [reply]
> With the font scale at 64,
> 
> For "P" I get advances = {39}. For "PA" I get advances = {37, 41}.

That shows some kerning happening. I get similar, though not exactly the same, with hb-shape:
```
behdad:build 0 (main)$ ~/hb/build/util/hb-shape ~/noto-fonts/unhinted/ttf/NotoSans/NotoSans-Regular.ttf PA --font-size 64
[P=0+36|A=1+41]
behdad:build 0 (main)$ ~/hb/build/util/hb-shape ~/noto-fonts/unhinted/ttf/NotoSans/NotoSans-Regular.ttf P --font-size 64
[P=0+39]
```
Using `hb-view` to render "PAH" I get:
![PAH](https://user-images.githubusercontent.com/142891/155608188-67b45eb7-43a6-4c76-8e1e-636ff95feed6.png)

I suspect your rendering code is not taking account of leftSideBearing when positioning glyphs. Are you using FreeType?

**Reply by behdad** [reply]
Check https://freetype.org/freetype2/docs/glyphs/glyphs-3.html

**Reply by bclever1** [reply]
Wow, that looks really nice. Yes, we are using FreeType. Also, using hb-view I get the same nice result.

**Reply by behdad** [reply]
> Wow, that looks really nice. Yes, we are using FreeType. Also, using hb-view I get the same nice result.

The bitmap you get from FreeType needs to be positioned correctly. That's all you need and you will get the same as hb-view.

**Reply by bclever1** [reply]
Aha I see! Thank you very much behdad, you have been more than patient and helpful.

**Reply by behdad** [reply]
:)

**Comment by behdad** [comment]
Do you have the `hb-shape` tool built, can you grab it from a HarfBuzz release? Does that return correct results for this font?
